src/Razor/src/Microsoft.CodeAnalysis.Razor.CohostingShared/Diagnostics/CohostDocumentPullDiagnosticsEndpointBase.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.Remote;
+using Microsoft.CodeAnalysis.Razor.Telemetry;
+using ExternalHandlers = Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost.Handlers;
+
+namespace Microsoft.VisualStudio.Razor.LanguageClient.Cohost;
+
+internal abstract class CohostDocumentPullDiagnosticsEndpointBase<TRequest, TResponse>(
+    IRemoteServiceInvoker remoteServiceInvoker,
+    IHtmlRequestInvoker requestInvoker,
+    ITelemetryReporter telemetryReporter,
+    ILoggerFactory loggerFactory)
+    : AbstractRazorCohostDocumentRequestHandler<TRequest, TResponse>
+    where TRequest : notnull
+{
+    private readonly IRemoteServiceInvoker _remoteServiceInvoker = remoteServiceInvoker;
+    private readonly IHtmlRequestInvoker _requestInvoker = requestInvoker;
+    private readonly ITelemetryReporter _telemetryReporter = telemetryReporter;
+    private readonly ILogger _logger = loggerFactory.GetOrCreateLogger<CohostDocumentPullDiagnosticsEndpointBase<TRequest, TResponse>>();
+
+    protected override bool MutatesSolutionState => false;
+
+    protected override bool RequiresLSPSolution => true;
+
+    protected abstract string LspMethodName { get; }
+    protected abstract bool SupportsHtmlDiagnostics { get; }
+
+    protected virtual LspDiagnostic[] ExtractHtmlDiagnostics(TResponse result)
+    {
+        throw new NotImplementedException("If SupportsHtmlDiagnostics is true, you must implement GetHtmlDiagnostics");
+    }
+
+    protected virtual TRequest CreateHtmlParams(Uri uri)
+    {
+        throw new NotImplementedException("If SupportsHtmlDiagnostics is true, you must implement CreateHtmlParams");
+    }
```

Comment:
💭 Nit: These technically shouldn't throw `NotImplementedException`. `NotSupportedException` or `InvalidOperationException` is more appropriate.